File: sdk/typescript/examples/externalSystemDemo.ts
import { CompliAGLClient, ExecutionResultStatus, attachResultSignature } from '../src';
import { startMockServer } from '../test/mockServer';

const usingExternal = Boolean(process.env.COMPLIAGL_BASE_URL);
const mock = usingExternal ? undefined : await startMockServer();
const client = new CompliAGLClient({
  baseUrl: process.env.COMPLIAGL_BASE_URL ?? mock!.baseUrl,
  organizationId: process.env.COMPLIAGL_ORG_ID ?? 'test-org',
  apiKey: process.env.COMPLIAGL_API_KEY ?? 'test-key',
});

try {
  console.log('CompliAGL lifecycle demo. External app: OUTSIDE CompliAGL generic fulfillment system.');
  const actor = await client.actorIdentities.create({ externalActorId: 'fulfillment-agent-001', actorType: 'AGENT', displayName: 'Outside Fulfillment Agent' });
  const target = await client.targets.create({ externalTargetId: 'supplier-001', targetType: 'MERCHANT', displayName: 'Outside Supplier' });
  const intent = await client.intents.create({ actorIdentityId: actor.id, action: 'PURCHASE', targetId: target.id, amount: 2500, currency: 'USD', payload: { item: 'standard-service' } });
  const context = await client.operationalContexts.create({ intentId: intent.id, targetId: target.id, environment: 'production', facts: { channel: 'demo' } });
  console.log('1 create actor/intent/target/context', actor.id, intent.id, target.id, context.id);
  const evaluation = await client.evaluations.create({ intentId: intent.id, operationalContextId: context.id });
  const resolved = await client.evaluations.resolve(evaluation.id);
  console.log('2 evaluation resolved', resolved.result);
  const source = await client.evidence.createSource({ sourceType: 'external-context', payload: { customerTier: 'standard' } });
  const collection = await client.evidence.createCollection({ evaluationId: evaluation.id, evidenceSourceIds: [source.id] });
  console.log('3 evidence collected', collection.id);
  const decision = await client.decisions.decide({ intentId: intent.id, evaluationId: evaluation.id, evidenceCollectionId: collection.id });
  console.log('4 decision read from CompliAGL', decision.outcome);
  const authorization = await client.authorizations.issue({ decisionId: decision.id });
  console.log('5 authorization issued', authorization.id, 'cap', authorization.approvedAmount);
  const authorizationCheck = await client.authorizations.verify(authorization.id);
  console.log('6 authorization verified', authorizationCheck.valid);
  console.log('7 OUTSIDE CompliAGL system executes generic fulfillment');
  const now = new Date().toISOString();
  const signed = attachResultSignature({ executionResultId: 'outside-result-100', authorizationId: authorization.id, externalSystemId: 'outside-fulfillment', status: ExecutionResultStatus.SUCCEEDED, executedAction: 'CAPTURE', target: { targetId: target.id, type: 'MERCHANT' }, amountMinor: 2500, amountCurrency: 'USD', externalReference: 'outside-order-100', paymentOrSettlementReference: 'settlement-100', executedAt: now, submittedAt: now, signerKeyId: 'demo-key-1', provenance: { system: 'outside-fulfillment', outsideCompliagl: true }, resultPayload: { fulfilled: true, capturedMinor: 2500 } }, 'result-secret');
  const result = await client.executionResults.submit(signed, { idempotencyKey: 'demo-result' });
  const binding = await client.verification.verifyExecutionResultBinding({ resultId: result.id });
  console.log('8 signed result submitted and binding valid', binding.valid);
  const proof = await client.aiproofs.generate({ resultId: result.id });
  const proofCheck = await client.aiproofs.verify(proof.id);
  console.log('9 AIProof generated and verified', proofCheck.verified, proof.id);
  console.log('Demo complete. CompliAGL authorized and proved; execution happened outside CompliAGL.');
} finally {
  await mock?.close();
}

File: sdk/python/examples/external_system_demo.py
from __future__ import annotations

import os
import pathlib
import sys
import uuid

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[1]))

from compliagl import CompliAGL, CompliAGLConfig, ExecutionResultStatus, ExternalExecutionResult, sign_execution_result


class SimulatedExternalCommerceSystem:
    """This adapter is outside CompliAGL; it performs the simulated real-world action."""

    def select_offer(self):
        return {"offer_id": "generic-offer-001", "description": "Generic service package", "amount_minor": 750, "currency": "USD"}

    def execute(self, authorization, params):
        return {
            "order_id": "order-" + uuid.uuid4().hex[:8],
            "fulfilled": True,
            "amount_minor": params["amount_minor"],
            "currency": params["currency"],
        }


def main():
    mock = None
    base_url = os.environ.get("COMPLIAGL_BASE_URL")
    if not base_url:
        from tests.mock_server import MockCompliAGLServer

        mock = MockCompliAGLServer().start()
        base_url = mock.base_url
        print(f"Using in-process mock CompliAGL server at {base_url}")
    try:
        org_id = os.environ.get("COMPLIAGL_ORG_ID", "org_demo")
        api_key = os.environ.get("COMPLIAGL_API_KEY")
        sdk = CompliAGL(CompliAGLConfig(base_url=base_url, organization_id=org_id, api_key=api_key, auto_idempotency=True, retry_base_ms=0))
        external = SimulatedExternalCommerceSystem()

        print("1. [EXTERNAL] OUTSIDE CompliAGL: select an offer")
        offer = external.select_offer()
        print("   ", offer)

        print("2. [SDK->CompliAGL] create actor identity, intent, target, and operational context")
        actor = sdk.actor_identities.create({"actor_type": "SERVICE", "credential_type": "NONE", "display_name": "External commerce app"})
        intent = sdk.intents.create({"actor_identity_id": actor["id"], "intent_type": "PAYMENT", "status": "SUBMITTED", "amount_minor": offer["amount_minor"], "amount_currency": offer["currency"]})
        target = sdk.targets.create({"target_type": "MERCHANT", "name": "Generic external marketplace", "external_id": offer["offer_id"]})
        context = sdk.operational_contexts.create({"environment": "TEST", "intent_id": intent["id"], "target_id": target["id"]})

        print("3. [SDK->CompliAGL] request evaluation, resolve it in mock, and read a server decision")
        evaluation = sdk.evaluations.create({"intent_id": intent["id"], "target_id": target["id"], "operational_context_id": context["id"]})
        sdk.evaluations.resolve(evaluation["id"], "APPROVED", ["MOCK_APPROVED"], "RESOLVED")
        decision = sdk.decisions.decide(org_id, evaluation["id"])
        print("   Decision outcome from CompliAGL:", decision["outcome"])

        print("4. [SDK->CompliAGL] issue signed execution authorization")
        authorization = sdk.authorizations.issue({"decision_id": decision["id"], "max_amount_minor": 1000, "amount_currency": "USD"})

        print("5. [SDK] verify authorization structure with CompliAGL")
        print("   ", sdk.verification.verify_authorization(authorization["id"], {"amount_currency": "USD"})["valid"])

        print("6. [EXTERNAL] OUTSIDE CompliAGL: perform the real action")
        outcome = external.execute(authorization, offer)
        print("   ", outcome)

        print("7. [SDK] sign and submit the ExternalExecutionResult")
        now = "2026-08-02T13:00:00Z"
        unsigned = ExternalExecutionResult(
            execution_result_id=str(uuid.uuid4()),
            authorization_id=authorization["id"],
            external_system_id="external-commerce-demo",
            status=ExecutionResultStatus.SUCCEEDED,
            executed_action="fulfill_generic_order",
            target={"offer_id": offer["offer_id"]},
            amount_minor=offer["amount_minor"],
            amount_currency=offer["currency"],
            external_reference=outcome["order_id"],
            payment_or_settlement_reference="settlement-" + outcome["order_id"],
            result_payload_hash="",
            executed_at=now,
            submitted_at=now,
            signer_key_id="",
            signature="",
            provenance={"system": "SimulatedExternalCommerceSystem", "outside_compliagl": True},
            result_payload=outcome,
        )
        result = sign_execution_result(unsigned, "demo-secret", "demo-key")
        submitted = sdk.execution_results.create(result)

        print("8. [CompliAGL] mock validates authorization/result binding on submit")
        print("   accepted result:", submitted["execution_result_id"])

        print("9. [SDK->CompliAGL] generate, retrieve, and verify AIProof")
        proof = sdk.aiproofs.generate(submitted["execution_result_id"])
        fetched = sdk.aiproofs.get(proof["id"])
        verified = sdk.aiproofs.verify(fetched["id"])
        print("   proof:", fetched["id"], "valid:", verified["valid"])
        print("Lifecycle complete. External execution happened OUTSIDE CompliAGL; CompliAGL authorized and proved it.")
    finally:
        if mock:
            mock.stop()


if __name__ == "__main__":
    main()
